Split flux communication and DG boundary scheme

nilsvu opened the pull request:
## Proposed changes

Decouple flux communication actions from DG boundary scheme

Separates the flux communication code into the actions that do the
communication and the DG boundary schemes that define what needs
to be communicated. Here's a breakdown of the changes:

- Delete FluxCommunicationTypes file

  This collection of types, templated on the Metavariables, tied all
  of the fluxcomm and DG code together. Removing it breaks up unnecessary
  dependencies between the low-level DG functions, the mid-level DataBox
  code and the high-level parallel actions.

- Add CollectDataForFluxes action

  Collects the data needed by the DG boundary scheme and stores it on
  mortars, projecting it if necessary.
  Collecting data on internal element interfaces allows them to be
  communicated in a subsequent action. Collecting data on external element
  boundaries relieves the actions that impose boundary conditions from
  this responsibility, thus removing their dependence on the DG scheme.
  Executables don't need to use this actions if they have another way
  of contributing data to mortars.

- Refactor flux communication actions

  Only send and receive the boundary data that is present on mortars,
  not responsible for computing it anymore.
  Also, only asynchronous boundary schemes, i.e. those where neighboring
  elements may not send and receive data at the same temporal IDs (e.g.
  for local time-stepping), need to keep track of the "next" temporal
  ID on mortars. The logic greatly simplifies for synchronous schemes.

- All numerical fluxes now conform to a protocol that defines and documents
  their compile-time interface.

- Executables define the DG boundary scheme, e.g. select between the
  standard and LTS strong first-order DG scheme. They may use multiple DG
  schemes, e.g. for different sets of variables as present in the linear
  and nonlinear elliptic solves, since the flux communication code is not
  templated on the Metavariables anymore but on the boundary scheme.

### Types of changes:

- [ ] Bugfix
- [ ] New feature
- [x] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

kidder commented on src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/FluxCommunication.hpp:
```diff
@@ -182,135 +229,105 @@ struct ReceiveDataForFluxes {
+            remote_data.at(direction).project_to_mortar(
+                face_meshes.at(direction), mortar_meshes.at(mortar_id),
+                mortar_sizes.at(mortar_id));
+        auto local_data_on_mortar = local_data.at(direction).project_to_mortar(
```
can we avoid projecting twice: it is very expensive

nilsvu commented:
@wthrowe I have refactored this code according to what we discussed in person. In particular:
- Because of conservation arguments we typically need the same data locally and remotely for computing numerical fluxes, so I don't distinguish between the data types anymore. This allows us to do projections only once per communication step.
- I factored the code that _collects_ the data to communicate out of the action that _sends_ it. So eventually we can wait with sending data that is local to the processor until we are done with computations on the element to use caches more efficiently.
- Collecting the data in a separate action also allows using the same action to collect "ghost" data to impose boundary conditions on external boundaries, so that solves another issue we talked about.

I have not adjusted any tests yet, and only the elliptic and the scalar wave executables so far (with and without LTS though). Could you take a look at this PR and review the architecture without worrying about implementation details, so I don't waste time implementing tests for things that may change? Then I implement tests and get this ready for a more thorough review once you give me the go-ahead. Does that work for you?

nilsvu commented:
@wthrowe could you take a look at this PR when you get the chance?

wthrowe commented on src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/CollectDataForFluxes.hpp:
```diff
@@ -0,0 +1,199 @@
+        // projecting, because in that case the face touches two mortars so we
+        // need to keep the data around.
+        auto boundary_data_on_mortar =
+            mortar_meshes.at(mortar_id) == face_meshes.at(direction)
```
You also need to check the mortar size, which is what your comment actually describes.  Look at the check at the bottom of MortarHelpers.hpp on develop.

Also in regards to the comment, you can require a projection when the face only touches one mortar.  It's just a resolution change in that case.

wthrowe commented on src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/FluxCommunication.hpp:
```diff
@@ -173,145 +179,89 @@ struct ReceiveDataForFluxes {
+        // neighbor
+        const auto& local_boundary_data_on_mortar =
+            all_mortar_data.at(mortar_id).local_data(temporal_id);
+        auto remote_boundary_data_on_mortar = local_boundary_data_on_mortar;
```
`local_boundary_data_on_mortar` does not appear to be used except to copy construct `remote_boundary_data_on_mortar`.  Just elide it.  What's going on is clear from the comment.

wthrowe commented on src/NumericalAlgorithms/DiscontinuousGalerkin/BoundarySchemes/StrongFirstOrder/StrongFirstOrder.hpp:
```diff
@@ -0,0 +1,272 @@
+          numerical_flux_computer,
+      const Mesh<Dim - 1>& face_mesh,
+      const db::const_item_type<NormalDotFluxesTag>& normal_dot_fluxes,
+      const db::item_type<
```
const_item_type

wthrowe commented on src/NumericalAlgorithms/DiscontinuousGalerkin/BoundarySchemes/StrongFirstOrder/StrongFirstOrder.hpp:
```diff
@@ -0,0 +1,272 @@
+  // Compute the Tags::NormalDotNumericalFlux from local and remote data
+  db::const_item_type<
+      db::add_tag_prefix<::Tags::NormalDotNumericalFlux, VariablesTag>>
+      normal_dot_numerical_fluxes(mortar_mesh.number_of_grid_points(), 0.0);
```
Is zero-initializing necessary?

wthrowe commented on src/NumericalAlgorithms/DiscontinuousGalerkin/BoundarySchemes/StrongFirstOrder/StrongFirstOrder.hpp:
```diff
@@ -0,0 +1,272 @@
+ * \ingroup DiscontinuousGalerkinGroup
+ * \brief Boundary contributions for a strong first-order DG scheme.
+ *
+ * \details Computes n.(F^*-F), lifts it to the volume and adds the contribution
```
TeX the math.

wthrowe commented on src/NumericalAlgorithms/DiscontinuousGalerkin/BoundarySchemes/StrongFirstOrder/StrongFirstOrderLocalTimeStepping.hpp:
```diff
@@ -0,0 +1,103 @@
+
+template <size_t Dim, typename VariablesTag, typename NumericalFluxComputerTag,
+          typename TemporalIdTag, typename TimeStepperTag>
+struct StrongFirstOrderLocalTimeStepping {
```
This needs some documentation, even if that is basically just a link to the GTS version.

wthrowe commented on src/Time/BoundaryHistory.hpp:
```diff
@@ -103,6 +104,18 @@ class BoundaryHistory {
   //@}
 
+  const LocalVars& local_data(const TimeStepId& time_id) const noexcept {
+    const Time& time = time_id.substep_time();
```
This will have a very mild conflict with a PR I intend to submit soon, which changes the internal time storage in this class from `Time` to `TimeStepId`.  That change should actually make this function more robust.  It should be an easy merge for whichever of us needs to handle it.

wthrowe reviewed (changes requested):
Overall, I think this is a good direction to go in.

nilsdeppe commented:
@wthrowe I think this is ready for you to take another look whenever you get a chance :)

nilsvu commented:
Yes please take a look at the fixups when you get the chance. However I haven‘t pinged you on this yet because more fixups are coming that add more tests.

nilsvu commented:
@wthrowe I finished writing all tests now, so this PR is feature-complete. To make reviewing for you easier, I pushed my changes as a lot of individual fixups on top of the commits you have already reviewed. Before getting a second reviewer involved, I intend to squash them together and then factor out as many self-contained commits as possible, perhaps even to separate PRs. Would you like me to do this now, or do you prefer to review the fixups first?

If there's anything to discuss about this PR, it's probably helpful to get into a hangout. Just ping me here or on email.

Note: All tests pass on clang-9, but there's a metaprogramming issue on gcc<=8 that I still need to fix.

nilsvu committed:
Decouple flux communication actions from DG boundary scheme

Separates the flux communication code into the actions that do the
communication and the DG boundary schemes that define what needs
to be communicated. Here's a breakdown of the changes:

- Delete FluxCommunicationTypes file

  This collection of types, templated on the Metavariables, tied all
  of the fluxcomm and DG code together. Removing it breaks up unnecessary
  dependencies between the low-level DG functions, the mid-level DataBox
  code and the high-level parallel actions.

- Add CollectDataForFluxes action

  Collects the data needed by the DG boundary scheme and stores it on
  mortars, projecting it if necessary.
  Collecting data on internal element interfaces allows them to be
  communicated in a subsequent action. Collecting data on external element
  boundaries relieves the actions that impose boundary conditions from
  this responsibility, thus removing their dependence on the DG scheme.
  Executables don't need to use this actions if they have another way
  of contributing data to mortars.

- Refactor flux communication actions

  Only send and receive the boundary data that is present on mortars,
  not responsible for computing it anymore.
  Also, only asynchronous boundary schemes, i.e. those where neighboring
  elements may not send and receive data at the same temporal IDs (e.g.
  for local time-stepping), need to keep track of the "next" temporal
  ID on mortars. The logic greatly simplifies for synchronous schemes.

- All numerical fluxes now conform to a protocol that defines and documents
  their compile-time interface.

- Executables define the DG boundary scheme, e.g. select between the
  standard and LTS strong first-order DG scheme. They may use multiple DG
  schemes, e.g. for different sets of variables as present in the linear
  and nonlinear elliptic solves, since the flux communication code is not
  templated on the Metavariables anymore but on the boundary scheme.

nilsvu committed:
Add has_received_from_all_mortars function

nilsvu committed:
Move flux communication actions to ParallelAlgorithms

nilsvu committed:
Add another numerical fluxes helper function overload

nilsvu committed:
Remove a disfunctional check from NumericalFlux protocol

We can't currently check that the package_data function is callable
because the `argument_tags` may contain base tags and we can't resolve
their types.

nilsvu commented:
Hey @wthrowe, I have rebased this PR on develop, so this is now the final refactor in this series of PRs. It's a fairly large diff, but I don't think I can split anything else off into independent commits. Instead, I split the PR into smaller commits that I will squash into one when review is complete to make it easier to review. Those commits are labeled `[squash]`. I hope this works for you. Please let me know if I can do more to help you review this PR.

Here are a few notes on individual commits:
- "Add CollectDataForFluxes action" depends on the other commits only weakly through the changes to the `InitializeMortars` action that's used in the test. To keep the test in the same commit as the tested code I also labeled it `[squash]`.
- "Move flux communication actions" together with "Flux communication actions" makes sure we keep the history of the fluxcomm action files.

wthrowe commented on tests/Unit/ParallelAlgorithms/DiscontinuousGalerkin/Test_CollectDataForFluxes.cpp:
```diff
@@ -0,0 +1,200 @@
+
+namespace {
+constexpr size_t VolumeDim = 2;
+constexpr size_t FaceDim = VolumeDim - 1;
```
`volume_dim` and `face_dim`.

nilsvu replied:
I get a compiler segfault when I make this change because line 62 defines a static variable with the same name in a class:

```cpp
static constexpr size_t volume_dim = volume_dim;
```

Should I bother resolving this, e.g. by templating the class on `VolumeDim` and passing `volume_dim` when instantiating it?

wthrowe replied:
Does `static constexpr size_t volume_dim = ::volume_dim;` work?

nilsvu replied:
yes of course, don't know how I could miss that 🤦‍♂️ let's see if an older compiler complains about shadowing. Edit: looks good.

wthrowe commented on tests/Unit/ParallelAlgorithms/DiscontinuousGalerkin/Test_CollectDataForFluxes.cpp:
```diff
@@ -0,0 +1,200 @@
+  bool is_projected;
+  TemporalId temporal_id;
+  size_t num_points;
+  BoundaryData project_to_mortar(
```
Test a case where projection is necessary.

wthrowe commented on src/ParallelAlgorithms/DiscontinuousGalerkin/HasReceivedFromAllMortars.hpp:
```diff
@@ -0,0 +1,54 @@
+/*!
+ * \ingroup DiscontinousGalerkinGroup
+ * \brief Determines if data on all mortars has been received for the `InboxTag`
+ * at time `temporal_id`.
```
This documentation doesn't seem to show up anywhere.

wthrowe commented on tests/Unit/ParallelAlgorithms/DiscontinuousGalerkin/CMakeLists.txt:
```diff
@@ -4,7 +4,7 @@
 set(LIBRARY "Test_ParallelDiscontinuousGalerkin")
 
 set(LIBRARY_SOURCES
-  Test_InitializeDomain.cpp
```
Why was this removed?

nilsvu replied:
something went wrong in a rebase, fixed

wthrowe commented on src/NumericalAlgorithms/DiscontinuousGalerkin/NumericalFluxes/NumericalFluxHelpers.hpp:
```diff
@@ -97,6 +97,30 @@ void normal_dot_numerical_fluxes(
+      numerical_flux_computer, packaged_data_int, packaged_data_ext,
+      typename NumericalFluxType::package_field_tags{},
+      typename NumericalFluxType::package_extra_tags{},
+      std::forward<gsl::not_null<NormalDotNumericalFluxes*>>(
```
`std::forward` doesn't really make sense if the reference type of the argument is known.  I think this works out to be the same as `std::move`, but that doesn't do anything anyway because move and copy are the same thing for `gsl::not_null`.  Just pass `n_dot_num_fluxes...`.

kidder replied:
ping

wthrowe commented on src/Elliptic/DiscontinuousGalerkin/NumericalFluxes/FirstOrderInternalPenalty.hpp:
```diff
@@ -168,8 +171,9 @@ struct FirstOrderInternalPenalty<Dim, FluxesComputerTag,
-    // `packaged_data` is a `Variables` which can only hold tensors.
+    // This will improve conditioning on non-uniform meshes. It can be
+    // implemented as of https://github.com/sxs-collaboration/spectre/pull/1725
+    // (and will be ASAP) by using the `package_extra_tags` below.
```
If you need a note for this file an issue.  Don't reference the current PR in a comment.

nilsvu replied:
removed the comment. I don't need an issue since this is my own todo, will implement this ASAP after this PR is merged (it's one of the features this PR makes possible).

wthrowe commented on src/ParallelAlgorithms/DiscontinuousGalerkin/InitializeMortars.hpp:
```diff
@@ -145,27 +134,30 @@ struct InitializeMortars {
+        std::move(box),
+        cpp17::negation<
+            std::is_same<typename BoundaryScheme::temporal_id_tag,
+                         typename BoundaryScheme::receive_temporal_id_tag>>{});
```
Either give this parameter a name below to explain its meaning or remove the `negation` and swap the argument types below.

wthrowe commented on tests/Unit/ParallelAlgorithms/DiscontinuousGalerkin/Test_InitializeMortars.cpp:
```diff
@@ -153,17 +147,17 @@ SPECTRE_TEST_CASE("Unit.ParallelDG.InitializeMortars", "[Unit][Actions]") {
-    CHECK(mortar_next_temporal_ids.at(interface_mortar_id) == 0);
+    // const auto& mortar_next_temporal_ids =
+    //     get_tag(Tags::Mortars<Tags::Next<TemporalIdTag>, 1>{});
+    // CHECK(mortar_next_temporal_ids.at(interface_mortar_id) == 0);
```
Enable or remove.

nilsvu replied:
added a fixup that tests this properly

wthrowe commented on tests/Unit/ParallelAlgorithms/DiscontinuousGalerkin/Test_InitializeMortars.cpp:
```diff
@@ -219,10 +213,10 @@ SPECTRE_TEST_CASE("Unit.ParallelDG.InitializeMortars", "[Unit][Actions]") {
+    // const auto& mortar_next_temporal_ids =
+    //     get_tag(Tags::Mortars<Tags::Next<TemporalIdTag>, 2>{});
+    // CHECK(mortar_next_temporal_ids.at(interface_mortar_id_east) == 0);
+    // CHECK(mortar_next_temporal_ids.at(interface_mortar_id_south) == 0);
```
Enable or remove.

wthrowe commented on tests/Unit/ParallelAlgorithms/DiscontinuousGalerkin/Test_InitializeMortars.cpp:
```diff
@@ -303,11 +297,11 @@ SPECTRE_TEST_CASE("Unit.ParallelDG.InitializeMortars", "[Unit][Actions]") {
+    //     get_tag(Tags::Mortars<Tags::Next<TemporalIdTag>, 3>{});
+    // CHECK(mortar_next_temporal_ids.at(interface_mortar_id_right) == 0);
+    // CHECK(mortar_next_temporal_ids.at(interface_mortar_id_front) == 0);
+    // CHECK(mortar_next_temporal_ids.at(interface_mortar_id_top) == 0);
```
Enable or remove.

wthrowe commented on src/ParallelAlgorithms/DiscontinuousGalerkin/FluxCommunication.hpp:
```diff
@@ -13,83 +10,282 @@
+ *
+ * Waits until remote boundary data from all internal mortars has arrived, then
+ * contributes it to the data in `Tags::Mortars<BoundaryScheme::mortar_data_tag,
+ * volume_dim>` that already holds the local boundary data.
```
This documentation should say "local time-stepping" or "asynchronous" or something along those lines somewhere.

wthrowe commented on src/ParallelAlgorithms/DiscontinuousGalerkin/FluxCommunication.hpp:
```diff
@@ -13,83 +10,282 @@
+ *
+ * For boundary schemes that communicate asynchronously (e.g. for local
+ * time-stepping) see the documentation of this action's template
+ * specialization.
```
It would be nice if this could be linked to, but I don't know if that's actually doable.

nilsvu replied:
I haven't manage to get that link working, but I added a link from the specialization back to here.

wthrowe commented on tests/Unit/ParallelAlgorithms/DiscontinuousGalerkin/Test_FluxCommunication.cpp:
```diff
@@ -604,249 +427,160 @@ void test_p_refinement() noexcept {
-       db::item_type<mortar_meshes_tag<3>>{{mortar_id, face_mesh}},
-       db::item_type<mortar_sizes_tag<3>>{{mortar_id, {{}}}}});
+      {initial_extents, time, mesh_neighbor,
+       // Unused
```
Make it clear what is unused.

wthrowe commented on src/NumericalAlgorithms/DiscontinuousGalerkin/Tags.hpp:
```diff
@@ -19,6 +19,12 @@
+/// Functionality related to discontinuous Galerkin schemes
+namespace dg {
+// The classes below should be moved in the `dg` namespace. See issue:
+// https://github.com/sxs-collaboration/spectre/issues/2140
```
We've got an issue, we don't need a code comment.

nilsvu replied:
I moved this commit to #2142, removed the comment there

wthrowe commented on tests/Unit/ParallelAlgorithms/DiscontinuousGalerkin/Test_InitializeMortars.cpp:
```diff
@@ -91,76 +94,108 @@ struct ElementArray {
+  const auto& mortar_data = get_tag(::Tags::Mortars<MortarDataTag, Dim>{});
+  for (const auto& mortar_id_and_mesh : expected_mortar_meshes) {
+    // Just make sure this exists, it is not expected to hold any data
+    mortar_data.at(mortar_id_and_mesh.first);
```
`CHECK(mortar_data.count(mortar_id_and_mesh.first) == 1)` is clearer and will give a better error message if it fails.

nilsvu replied:
Thanks for the suggestion! I changed it to `CHECK(mortar_data.find(mortar_id_and_mesh.first) != mortar_data.end())`, how's that? Can change to `count` if you prefer

wthrowe replied:
Yes, that's fine.  As long as it's a CHECK.

wthrowe commented on src/ParallelAlgorithms/DiscontinuousGalerkin/FluxCommunication.hpp:
```diff
@@ -0,0 +1,379 @@
+ * This template specialization handles asynchronous boundary schemes, e.g.
+ * for local time-stepping. It proceeds only once the received boundary data
+ * on every mortar has caught up with the `BoundaryScheme::receive_temporal_id`
+ * on that mortar.
```
Not quite.  The received boundary data has to catch up with the BoundaryScheme::receive_temporal_id on the element, not the mortar.  The one on the mortar represents when the next message is expected from that neighbor, which is the same as the value that that neighbor has for itself.  (Although these can't be updated until the action runs, so the is_ready function has to compute the value that they will have after the inbox is processed).

nilsvu commented:
@wthrowe rebased to resolve conflicts with #2135 and pushed fixups for your comments. Please take a look :)

wthrowe commented:
Looks good.  I'm the only recent reviewer, so I'll say you can squash the fixups.

nilsvu commented:
Thanks for reviewing @wthrowe! @kidder @nilsdeppe could one of you provide the second review?

nilsvu commented:
FYI, I'll hold off rebasing this until #2114 is merged since the CI is broken until then.

kidder commented:
I am reviewing this now, please don't push anything until the review appears, thanks!

kidder commented on src/ParallelAlgorithms/DiscontinuousGalerkin/HasReceivedFromAllMortars.hpp:
```diff
@@ -0,0 +1,54 @@
+bool has_received_from_all_mortars(
+    const TemporalIdType& temporal_id, const Element<Dim>& element,
+    const tuples::TaggedTuple<InboxTags...>& inboxes) noexcept {
+  if (element.number_of_neighbors() == 0) {
```
UNLIKELY

kidder commented on src/ParallelAlgorithms/DiscontinuousGalerkin/CollectDataForFluxes.hpp:
```diff
@@ -0,0 +1,203 @@
+ * the amount of data being communicated. This assumption allows us to perform
+ * only one projection to the mortar in each step, instead of projecting local
+ * and remote data separately.
+ * - Wording: "Boundary data" is data collected on an element interface
```
Wording -> Terminology

kidder commented on src/ParallelAlgorithms/DiscontinuousGalerkin/FluxCommunication.hpp:
```diff
@@ -13,83 +10,289 @@
+      const Parallel::ConstGlobalCache<Metavariables>& /*cache*/,
+      const ArrayIndex& /*array_index*/, const ActionList /*meta*/,
+      const ParallelComponent* const /*meta*/) noexcept {
+    if (get<domain::Tags::Element<volume_dim>>(box).number_of_neighbors() ==
```
UNLIKELY

kidder commented on src/NumericalAlgorithms/DiscontinuousGalerkin/NumericalFluxes/NumericalFluxHelpers.hpp:
```diff
@@ -117,9 +117,7 @@ void normal_dot_numerical_fluxes(
-      typename NumericalFluxType::package_extra_tags{},
-      std::forward<gsl::not_null<NormalDotNumericalFluxes*>>(
-          n_dot_num_fluxes)...);
+      typename NumericalFluxType::package_extra_tags{}, n_dot_num_fluxes...);
```
should this be in the commit  "Add another numerical fluxes helper function overload"

nilsvu replied:
yes, fixed

kidder commented on src/ParallelAlgorithms/DiscontinuousGalerkin/InitializeMortars.hpp:
```diff
@@ -48,49 +47,43 @@ namespace Actions {
- *   - Tags::Mortars<Tags::Next<temporal_id_tag>, dim>
- *   - Tags::Mortars<Tags::Mesh<dim - 1>, dim>
- *   - Tags::Mortars<Tags::MortarSize<dim - 1>, dim>
+ *   - `Tags::Mortars<BoundaryScheme::mortar_data_tag, dim>
```
missing a tick at the end

kidder commented on src/ParallelAlgorithms/DiscontinuousGalerkin/InitializeMortars.hpp:
```diff
@@ -145,27 +134,30 @@ struct InitializeMortars {
     }
 
+    // We don't need to initialize the "next" temporal id on mortars if data is
+    // communicated synchronously
```
please clarify what you mean, communication is always asynchronous in charm++

kidder commented on src/ParallelAlgorithms/DiscontinuousGalerkin/InitializeMortars.hpp:
```diff
@@ -177,6 +169,39 @@ struct InitializeMortars {
+ private:
+  template <typename DbTagsList>
+  static auto initialize_next_temporal_id_on_mortars(
+      db::DataBox<DbTagsList>&& box, std::true_type /* meta */) noexcept {
```
instead of meta can you replaces with is_bla. (where bla is whatever is true...)

nilsvu replied:
named `enabled` and `disabled`, since the function name already describes what is being enabled or disabled

kidder commented on src/ParallelAlgorithms/DiscontinuousGalerkin/InitializeMortars.hpp:
```diff
@@ -177,6 +169,39 @@ struct InitializeMortars {
+  template <typename DbTagsList>
+  static db::DataBox<DbTagsList> initialize_next_temporal_id_on_mortars(
+      db::DataBox<DbTagsList>&& box, std::false_type /* meta */) noexcept {
+    return std::forward<db::DataBox<DbTagsList>>(box);
```
just return box

nilsvu replied:
> warning: local variable 'box' will be copied despite being returned by name [-Wreturn-std-move]
> note: call 'std::move' explicitly to avoid copying

So I made this `std::move(box)`. Is that the correct way to forward the argument?

kidder commented on tests/Unit/ParallelAlgorithms/DiscontinuousGalerkin/Test_InitializeMortars.cpp:
```diff
@@ -57,118 +72,135 @@ struct ElementArray {
+void test_next_temporal_ids(
+    GetTag&& get_tag,
+    const dg::MortarMap<Dim, Mesh<Dim - 1>>& expected_mortar_meshes,
+    std::true_type /* meta */) {
```
clarify meta with what is true

kidder commented on tests/Unit/ParallelAlgorithms/DiscontinuousGalerkin/Test_InitializeMortars.cpp:
```diff
@@ -57,118 +72,135 @@ struct ElementArray {
+void test_next_temporal_ids(
+    GetTag&& /* get_tag */,
+    const dg::MortarMap<Dim, Mesh<Dim - 1>>& /* expected_mortar_meshes */,
+    std::false_type /* meta */) {}
```
and what is not true

kidder reviewed (changes requested):
reviewed with @nilsdeppe (please don't squash yet)

kidder commented:
@wthrowe did you want to look at the new fixups for this PR before it is squashed?

nilsvu commented:
To make sure I understand, should I squash the fixups only for now or also all commits labeled `[squash]`?

nilsdeppe commented:
don't squash anything for now

wthrowe commented:
The fixups look fine.

kidder commented on src/ParallelAlgorithms/DiscontinuousGalerkin/InitializeMortars.hpp:
```diff
@@ -173,7 +173,7 @@ struct InitializeMortars {
   template <typename DbTagsList>
   static auto initialize_next_temporal_id_on_mortars(
-      db::DataBox<DbTagsList>&& box, std::true_type /* meta */) noexcept {
+      db::DataBox<DbTagsList>&& box, std::true_type /* enabled */) noexcept {
```
we still do not know what is enabled. 
how about `perform_initialization`

kidder commented on src/ParallelAlgorithms/DiscontinuousGalerkin/InitializeMortars.hpp:
```diff
@@ -199,8 +199,8 @@ struct InitializeMortars {
   static db::DataBox<DbTagsList> initialize_next_temporal_id_on_mortars(
-      db::DataBox<DbTagsList>&& box, std::false_type /* meta */) noexcept {
-    return std::forward<db::DataBox<DbTagsList>>(box);
+      db::DataBox<DbTagsList>&& box, std::false_type /* disabled */) noexcept {
```
the comment should be the same in both sites (i.e. it is true above and false here)

kidder commented on tests/Unit/ParallelAlgorithms/DiscontinuousGalerkin/Test_InitializeMortars.cpp:
```diff
@@ -115,7 +115,7 @@ template <size_t Dim, typename GetTag>
     GetTag&& get_tag,
     const dg::MortarMap<Dim, Mesh<Dim - 1>>& expected_mortar_meshes,
-    std::true_type /* meta */) {
+    std::true_type /* asynchronous */) {
```
here (and below) `times_are_asynchronus`

kidder replied:
@nilsleiffischer sorry, my fault, please squash, `times_are_asynchronous`

nilsvu replied:
done

kidder reviewed (changes requested):
@nilsdeppe and I reviewed the fixups, and now you may squash as much as you like

nilsvu commented:
Ok I rebased and squashed everything. Thanks for reviewing and getting this refactor done :)

kidder commented:
@wthrowe this should be ready for your approval

nilsvu commented:
Rebased on merged #2120 to resolve a conflict. @wthrowe please take a look when you get the chance

nilsvu commented:
@kidder this looks ready to merge. Could you take a look?

nilsvu commented:
Rebased on merged #2193 to fix a conflict
